### PR TITLE
[DP-2212] Prevent redactions in {{placeholders}}

### PR DESCRIPTION
**Pull request by ag4ta**
Adds a new `protect_placeholders` option (default: false) to prevent the sanitizer from redacting numbers within placeholders like `{{ticket.ticket_field_4418815038484}}`.

 By `default protect_placeholders` is disabled, so there is no need to do anything after bumping the gem to the newer version if this feature is not needed.

Introduces a new protect_placeholders option that:
- Detects when a number sequence is within a placeholder using `{{` and `}}` delimiters (by default)
 - Skips sanitization of numbers inside valid placeholders
 - Still sanitizes real credit card numbers between or outside placeholders (`{{first}} 4418815038484 {{second}}`)
 - Handles edge cases like malformed placeholders (unclosed {{, nested delimiters, etc.)

# Manual tests
1. I created the Ticket's field called "Special Order Number" with "proper" Field ID that looks like credit card (in rails console using SQL insert :blush: )
image

2. Then added it to the Default Ticket Form:
image

 
3. And at the end, created the ticket, with two comments:
- first when the `protect_placeholders_in_credit_card_sanitization` feature flag was disabled
- second when the `protect_placeholders_in_credit_card_sanitization` feature flag was enabled

#### THE MESSAGE I USED:
```
New/Old behaviour:
4418815038484 => {{ticket.ticket_field_4418815038484}}
```

image

### CODE REVIEW

File: lib/credit_card_sanitizer.rb [hunk 1]
```diff
@@ -211,6 +219,29 @@ def valid_flanking_by_no_space_languages?(prefix, postfix)
...
+    last_open_index = prefix.rindex(PLACEHOLDER_OPEN)
+    last_close_index = prefix.rindex(PLACEHOLDER_CLOSE)
+
+    # Check if there's an unclosed opening delimiter
+    has_unclosed_opening = last_open_index && (last_close_index.nil? || last_open_index > last_close_index)
+    return false unless has_unclosed_opening
+
+    # Check if postfix has a closing delimiter
+    first_close_in_postfix = postfix.index(PLACEHOLDER_CLOSE)
+    return false unless first_close_in_postfix
```
Comment (ChrisGadek1) [inline lib/credit_card_sanitizer.rb:242]: Can we re-use here `last_close_index` variable? Not sure what's the difference

Image

**Review by ChrisGadek1 — APPROVED** [review]
Looks good, left one, small comment

File: lib/credit_card_sanitizer.rb (hunk 1 above)
Comment (ag4ta) [reply lib/credit_card_sanitizer.rb:242]: we can't re-use `last_close_index` as it comes from the `prefix` part, not `postfix`. 

It has this complexity in order to cover cases when a credit card is not part of any placeholder but it's surrounded by them:
```
{{first}} 4111111111111111 {{second}}  -> {{first}} 411111▇▇▇▇▇▇1111 {{second}}
```
in this case we have:
- the `prefix` is `{{first}}`
- the `postfix` is `{{second}}`





Additionally the `postfix` sometimes may look like this:
```
}} Card 5555555555554444 {{ticket.ticket_field_4418815038484}}
```
So we need to handle/cover such cases too :)

**Review by HeyNonster — APPROVED** [review]

File: lib/credit_card_sanitizer.rb [hunk 2]
```diff
@@ -57,6 +57,8 @@ class CreditCardSanitizer
   NUMBERS_WITH_LINE_NOISE = /#{SCHEME_OR_PLUS}?\d(?:#{LINE_NOISE}\d){10,30}/
   # Languages that don't use spaces between words/characters and numbers
   ACCEPTED_JAPANESE_CHINESE_CHARS = /[\u3040-\u309F\u30A0-\u30FF\u4E00-\u9FFF]/  # Japanese: Hiragana (あ-ゟ), Katakana (ア-ヿ), Kanji (一-鿿); Chinese: Hanzi (CJK Unified Ideographs)
+  PLACEHOLDER_OPEN = "{{"
+  PLACEHOLDER_CLOSE = "}}"
```
Comment (bquorning) [inline lib/credit_card_sanitizer.rb]: This being an open source project, I think we should consider making the placeholders configurable instead of hardcoded?

File: README.md [hunk 3]
```diff
@@ -43,6 +43,7 @@ is also returned.
 | `exclude_tracking_numbers`          | Identify shipping tracking numbers and don't truncate them. The default is `false`.                |
 | `parse_flanking`                    | Only sanitize credit card numbers with valid prefixes/postfixes. The default is `false`.           |
 | `allow_flanking_by_no_space_languages`| Allow sanitization of credit cards flanked by Japanese/Chinese characters. The default is `false`. |
+| `protect_placeholders`              | Do not sanitize numbers within placeholders like `{{...}}`. The default is `false`.        |
```
Comment (bquorning) [inline README.md]:
```suggestion
| `protect_placeholders`              | Do not sanitize numbers within placeholders like `{{...}}`. The default is `false`.                |
```

File: lib/credit_card_sanitizer.rb [hunk 4]
```diff
@@ -211,6 +219,29 @@ def valid_flanking_by_no_space_languages?(prefix, postfix)
     (prefix && ACCEPTED_JAPANESE_CHINESE_CHARS.match(prefix[-1])) || (postfix && ACCEPTED_JAPANESE_CHINESE_CHARS.match(postfix[0]))
   end
 
+  # Checks if the number is within a placeholder like {{...}}
+  def within_placeholder?(prefix, postfix)
+    last_open_index = prefix.rindex(PLACEHOLDER_OPEN)
```
Comment (bquorning) [inline lib/credit_card_sanitizer.rb]: You can safe an `rindex` operation by doing an early return after this line:
```suggestion
    last_open_index = prefix.rindex(PLACEHOLDER_OPEN)
    return false unless last_open_index
```

**Review by bquorning — COMMENTED** [review]
I am curious about the performance impact of this PR. Could you do a small benchmark of before/after with/without content in placeholders?

File: README.md (hunk 3 above)
Comment (ag4ta) [reply README.md]: fixed, the table is properly formatted now

File: lib/credit_card_sanitizer.rb (hunk 4 above)
Comment (ag4ta) [reply lib/credit_card_sanitizer.rb]: Thanks! I added it to my changes too!

File: lib/credit_card_sanitizer.rb (hunk 2 above)
Comment (ag4ta) [reply lib/credit_card_sanitizer.rb]: Good point! I introduced the `placeholder_open` and `placeholder_close` options (default: `{{` and `}}`)

File: test/credit_card_sanitizer_test.rb [hunk 5]
```diff
@@ -195,6 +195,85 @@
...
+          assert_nil @sanitizer.sanitize!("Process ${id} for card ${field_378282246310005}")
+        end
+
+        it "sanitizes when custom delimiters don't match" do
+          @sanitizer = CreditCardSanitizer.new(protect_placeholders: true, placeholder_open: "<%", placeholder_close: "%>")
+          # These use {{ }} but we configured <% %>, so they should be sanitized
+          assert_equal "{{ticket.ticket_field_441881▇▇▇8484}}", @sanitizer.sanitize!("{{ticket.ticket_field_4418815038484}}")
+        end
+
+        it "handles empty or nil placeholder delimiters" do
```
Comment (bquorning) [inline test/credit_card_sanitizer_test.rb]: I think you should actually test with nil placeholder delimiters too. I am not sure NilClass responds to `#empty?`.

**Comment by ag4ta** [comment]
> I am curious about the performance impact of this PR. Could you do a small benchmark of before/after with/without content in placeholders?

@bquorning, I asked Claude to perform for me a benchmark for messages that have 60KB with and without `protect_placeholders ` enabled 😅

Here are results:
```
✓ Realistic Scenario 
Example:
  Support ticket #123. Customer card is 4111 1111 1111 1111 for payment.
  Order {{ticket.ticket_field_4418815038484}} for {{requester.name}}.

  Content: 60KB with 4 credit cards
  Result: -0.096ms (2% FASTER)


⚠ Worst Case (Unrealistic)
Example:
  Card 1: 4111 1111 1111 1111, Card 2: 5555 5555 5555 4444,
  Card 3: 3782 822463 10005... (309 cards total!)

  Content: 60KB with 309 credit cards
  Result: +3.607ms (+55% slower)


✓✓ Best Case (Cards Protected by Placeholders)
Example:
  Payment card {{ticket.ticket_field_4418815038484}} value {{4111 1111 1111 1111}} processed.
  Another card {{ticket.ticket_field_9182736450011}} has {{5555 5555 5555 4444}} saved.

  Content: 60KB with 41 credit cards **in placeholders**
  Result: -0.879ms (15% FASTER!)


✓✓ Mixed (Most Common Real-World)
Example:
  Card 4111 1111 1111 1111 on file.
  Template {{ticket.ticket_field_4418815038484}} with {{5555 5555 5555 4444}} for user {{ticket.requester_id}}.

  Content: 60KB with 21 unprotected + 18 protected cards
  Result: -0.256ms (5% FASTER)
 
```
it's unlikely to have a 60KB message with only credit cards within it, and the benchmarks highly depends on given input/message.
So in general it depends on the message structure itself: whether it contains credit cards, how many of them etc

File: test/credit_card_sanitizer_test.rb (hunk 5 above)
Comment (ag4ta) [reply test/credit_card_sanitizer_test.rb]: ah yes! NilClass does not respond to `#empty?`
I'll fix it

Comment (ag4ta) [reply test/credit_card_sanitizer_test.rb]: Done 😊

**Review by bquorning — APPROVED** [review]
Yeah I agree, if anyone is sending 60KB of credit card numbers in an HTTP request we may need to have a talk with them…
